pkg/erofs/erofs_unsafe.go

```diff
@@ -0,0 +1,57 @@
+// Copyright 2023 The gVisor Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package erofs
+
+import (
+	"unsafe"
+
+	"gvisor.dev/gvisor/pkg/errors/linuxerr"
+)
+
+func (i *Image) pointerAt(off uint64) unsafe.Pointer {
+	return unsafe.Pointer(uintptr(unsafe.Pointer(&i.bytes[0])) + uintptr(off))
+}
+
+type scalar interface {
+	uint8 | uint16 | uint32 | uint64
+}
+
+// scalarAt returns the scalar value located at the specified position in the
+// memory backed by image.
+func scalarAt[T scalar](i *Image, off uint64) (T, error) {
+	var zero T
+	n := uint64(unsafe.Sizeof(zero))
+	if ok := i.checkRange(off, n); !ok {
+		return zero, linuxerr.EFAULT
+	}
+	return *(*T)(i.pointerAt(off)), nil
+}
+
+// Dirent cannot be used with objectAt, because Go does not support packed struct,
+// and its size in memory is different with its size on disk.
+type object interface {
+	SuperBlock | InodeCompact | InodeExtended
+}
```

Comment:
Hmm can't we just template `objectAt` with `any`? Does this constraint help? like all these types are Marshallable, so does the constraint allow us to call `SizeBytes()`?

Reply:
Yeah, I think it helps. Because we cannot use `objectAt[Dirent](...)` (because `Dirent` is a packed type on disk, its size in memory is different with its size on disk), and it will be better if compiler can raise an error in this case.

Reply:
Makes sense.